### ISSUE TITLE
internal/restorable: intense allocations at `resolveStale` in Android

### ISSUE THREAD

**Issue by corfe83** [hajimehoshi/ebiten#2375]
I'm noticing that the android port of my game is sometimes choppy. It only seems to be a problem when the garbage collector is running too frequently -- sometimes it runs almost constantly (every 50-100 milliseconds or so, other times it goes much longer). So I did some profiling, and found that ebiten EndFrame() is doing 98.77% of ALL the allocs in my entire Idle Armada game on Android! This is ONLY the case on the android port, windows is behaving fine.

How can we reduce allocs in atlas code?

![image](https://user-images.githubusercontent.com/61067263/194173432-55797cf5-9ca2-4fcd-aa10-a5c5639ee928.png)
[allocs.zip](https://github.com/hajimehoshi/ebiten/files/9719309/allocs.zip)

**Comment by corfe83** [comment]
This is not a complete or correct solution, but this simple change dramatically reduces number of allocations here. I didn't propose it as a pull request because I don't think it's properly thought out.

Is there a better generic solution to avoid having ebiten allocate so often? I think sync.Pool offers potentially interesting solutions to this problem, like this:

https://github.com/corfe83/ebiten/commit/dd33f5f0405c1e9c3012a5f0e1197f04f204a369

**Comment by hajimehoshi** [comment]
Actually allocation happens on Android in some situations. I think your suggestion makes sense. Let me think more.

> Is there a better generic solution to avoid having ebiten allocate so often? I think sync.Pool offers potentially interesting solutions to this problem, like this:

I think you are doing relatively complex rendering. These advices would be useful:

https://ebitengine.org/en/documents/performancetips.html#Make_similar_draw_function_calls_successive
https://ebitengine.org/en/documents/performancetips.html#Avoid_changing_render_sources'_pixels
https://ebitengine.org/en/documents/performancetips.html#Avoid_using_the_screen_as_a_render_source

**Comment by hajimehoshi** [comment]
I believe we can reuse `internal/atlas.temporaryBytes`

**Comment by hajimehoshi** [comment]
Could you try

```
go get github.com/hajimehoshi/ebiten/v2@8cf3c31cf6da12f03576a58bc5e7b4fc7c0ba708
```

? Thanks,

(As this is just a performance improvement, I don't plan to backport this to v2.4.)

**Comment by hajimehoshi** [comment]
By the way, how did you take a profile for Android?

**Comment by corfe83** [comment]
This is my profiling "utility" file -- then I use USB file transfer. I call updateProfiling() every frame, this makes sure it saves every 60 seconds.

`package engine

import (
	"os"
	"runtime"
	"runtime/pprof"
	"time"

	"github.com/corfe83/goutil/androidutil"
)

var profileTypes = [...]string{
	"goroutine",
	"heap",
	"allocs",
	"threadcreate",
	"block",
	"mutex",
}

var ProfilingEnabled bool = false
var perfDirectory string

func beginProfiling() {
	if ProfilingEnabled {
		if runtime.GOOS == "windows" || runtime.GOOS == "linux" {
			perfDirectory = "perf"
		} else if runtime.GOOS == "android" {
			perfDirectory = androidutil.AndroidSaveDataPath() + "perf"
		}

		if _, err := os.Stat(perfDirectory); os.IsNotExist(err) {
			os.Mkdir(perfDirectory, 0777)
		}

		beginCpuProfile()
		lastSaveTime = time.Now().Unix()
	}
}

var lastSaveTime int64

func updateProfiling() {
	if !ProfilingEnabled {
		return
	}

	if newTime := time.Now().Unix(); newTime >= lastSaveTime+60 {
		endProfiling()
		lastSaveTime = newTime
	}
}

func endProfiling() {
	if ProfilingEnabled {
		writeProfilingData()

		endCpuProfile()
		ProfilingEnabled = false
	}
}

func writeProfilingData() {
	if ProfilingEnabled {
		for _, profileType := range profileTypes {
			writeProfileFile(profileType)
		}
	}
}

func writeProfileFile(profileType string) {
	filePath := perfDirectory + string(os.PathSeparator) + profileType + ".prof"
	tempFilePath := filePath + "temp"
	f, err := os.Create(tempFilePath)
	if err != nil {
		panic(err)
	}
	runtime.GC() // get up-to-date statistics
	if err := pprof.Lookup(profileType).WriteTo(f, 0); err != nil {
		panic(err)
	}
	f.Close()

	if err := os.Rename(tempFilePath, filePath); err != nil {
		panic(err)
	}
}

func beginCpuProfile() {
	filePath := perfDirectory + string(os.PathSeparator) + "CPU.prof"
	f, err := os.Create(filePath)
	if err != nil {
		panic(err)
	}

	pprof.StartCPUProfile(f)
}

func endCpuProfile() {
	pprof.StopCPUProfile()
}`

**Comment by corfe83** [comment]
Thanks, I am trying the commit, and will look to make sure my draw calls are similar, and try ebitenginedebug build tag as well

**Comment by hajimehoshi** [comment]
> `go get github.com/hajimehoshi/ebiten/v2@8cf3c31cf6da12f03576a58bc5e7b4fc7c0ba708`

I'm afraid this might not work well (when an image becomes 'stale', all the pixels records are gone, then a byte slice cannot be reused). If this doesn't work, I'll try another solution that is close to your suggestion.

**Comment by corfe83** [comment]
I can confirm the commit does improve things a lot, but too bad if it won't work in all scenarios. I will investigate how wasteful my drawing code is, and would love to see some kind of improvement as well.

**Comment by hajimehoshi** [comment]
Oh, did the commit improve the situation? I thought the commit didn't improve `readPixelsFromGPU` at `resolveStale`, but was this changed?

**Comment by corfe83** [comment]
I didn't use a stopwatch, so the time of profiling is not exactly the same, but it was a similar timeframe, and with the commit it's a lot better on android. 327 MB instead of 12 GB of allocations:

![image](https://user-images.githubusercontent.com/61067263/194214558-98f3b0e8-149b-4c1e-801e-01222ea272d0.png)

**Comment by hajimehoshi** [comment]
> `getByteSliceFromPool`

Isn't this your commit rather than my commit `8cf3c31cf6da12f03576a58bc5e7b4fc7c0ba708`?

**Comment by corfe83** [comment]
You're right, apparently my test did not write a new profile file, and I copied the old one. I will try again.

**Comment by corfe83** [comment]
Does ebitenmobile need to be fixed in 2.5? Or is this my error?

```
            Ebitenmobileview.onGamepadAxesOrHatsChanged(event.getDeviceId(), axis, value);
                            ^
  symbol:   method onGamepadAxesOrHatsChanged(int,int,float)
  location: class Ebitenmobileview
com\musicalbox\idle\armada\mobile\EbitenView.java:227: error: method onGamepadAdded in class Ebitenmobileview cannot be applied to given types;
        Ebitenmobileview.onGamepadAdded(deviceId, inputDevice.getName(), nbuttons, naxes, nhats2/2, descriptor, vendorId, productId, buttonMask, axisMask);
                        ^
  required: long,String,long,long,String,long,long,long,long
  found: int,String,int,int,int,String,int,int,int,int
  reason: actual and formal argument lists differ in length
```

**Comment by hajimehoshi** [comment]
ebitenmobile for v2.4 and v2.5 are not compatible, so I recommend running ebitenmobile with `go run`

```
go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile # under your go.mod
```

**Comment by corfe83** [comment]
Unfortunately I think you're right, the change doesn't help:

![image](https://user-images.githubusercontent.com/61067263/194220250-fa826ded-cb41-4ac4-9732-b689ddb849e4.png)

I tried with debugging, and attached are my draw commands for one of the problematic screens. I can try to reduce drawing calls, but I'm not sure if this looks incredibly wasteful what I'm doing or not.

[ebitenginedebug.txt](https://github.com/hajimehoshi/ebiten/files/9721451/ebitenginedebug.txt)

**Comment by hajimehoshi** [comment]
Thank you for confirming!

> ebitenginedebug.txt

The number of draw calls seems sane, but the number of internal seems too big. Perhaps are you calling `NewImage*` every frame?

**Comment by corfe83** [comment]
[log.txt](https://github.com/hajimehoshi/ebiten/files/9721479/log.txt)
Here is a longer log. I do create new images on some frames, but I don't think it's a large number of frames that have this, I believe it happens when certain text controls update, my infrastructure creates an image for the rendered control with the new text.

I guess I'm not calling Dispose when I should, but it doesn't seem to be a big problem.

**Comment by corfe83** [comment]
I do keep a rather large number of small images in memory. I'm not sure if that's likely to be a problem.

**Comment by hajimehoshi** [comment]
My guess is that you created a lot of images and use them render targets. (A render tanget cannot be on a texture atlas)

**Comment by corfe83** [comment]
Yes, you're right I haven't disposed. I'm going to add Dispose() calls now, I somehow missed it when implementing my basic infrastructure (when I first started using ebiten).

In any case, the new commit (e66bac5c3f) is a huge improvement on alloc and GC activity for me! I'm unfortunately still getting some framerate issues, but it's not as bad, and it's clearly unrelated to GC now.

**Comment by hajimehoshi** [comment]
OK, let me close this issue as the memory/GC issue was resolved!

It's ok to continue discussion about your application's optimization here, or if you find a specific issue, please create a new file.

Thank you for reporting this and cooperating to confirm the fix!

**Comment by corfe83** [comment]
I see the original fix has been reverted in 2.5: 09e03203092773bb4a3762a04ab6da1674f9e6de, and I'm again getting huge allocs on android. I'm not sure if it was the right answer for my problems or not, so let me go into some detail about what I'm doing.

I'm trying to decide what to do to make things fast. I have a "Label" struct that I use over and over in my code for text and buttons. The UI looks like this -- each different button, label with a colored background, or white text with no background here is a separate image that I used as a draw target to construct it (once), and then I render to the screen (many times). There are probably 100 of these on screen at a time, and maybe 1000 in memory. Since each label/button has rounded corners, internally it is rendered as 7 images (4 corners + 3 rectangles), and then 1 text draw call on top. Rather than make 7 + 1 draw calls per label per frame, I render the 7 + 1 draw calls to a "cached" in-memory image, which I keep in memory until the button text or color changes (which is rare). When rendering each frame, I render the contents of the "cached" image to the screen. Remember there might be 100 of these drawn at once, and 1000 in memory at once.

![image](https://user-images.githubusercontent.com/61067263/223592259-7d88377e-ad75-4159-930f-0a371a340543.png)

All of this runs smoothly on windows/linux/mac, but on android (a fast phone, pixel 6 pro) there is a lot of lag and frame skipping.

I don't think I'm breaking any of the perf rules documented here, I don't do weird color things, I mostly just use GeoM:
https://ebitengine.org/en/documents/performancetips.html#Make_similar_draw_function_calls_successive

Still, 30% of CPU on android is used by ResolveStaleImages: 
![image](https://user-images.githubusercontent.com/61067263/223592699-1080ff52-01e3-4ecd-aabc-036d78bc27e5.png)

And an enormous amount of memory allocations are happening in ResolveStaleImages as well:
![image](https://user-images.githubusercontent.com/61067263/223592854-11fbf675-6263-49cf-8b95-1df3e3c27d0c.png)

Questions:
1) Is it bad to have 1000+ generated images kept in memory for ebiten? Should I avoid this and just make 7 + 1 draw calls directly to the screen for each label or button on screen?
2) Right now each codepath creates its own DrawImageOptions struct. Am I supposed to be passing the same one around to reuse for performance?

Attached is the Label struct I use. Not the highest quality code, but it's been useful for me. See the "caching" functionality I'm talking about in updateCachedImage(). Draw() is the function that writes the cached image to the screen.
[label.txt](https://github.com/hajimehoshi/ebiten/files/10915471/label.txt)

**Comment by corfe83** [comment]
Forgot to mention you @hajimehoshi , for visibility

**Comment by hajimehoshi** [comment]
Oops, this is an unintended regression. I'll take a look asap.

**Comment by corfe83** [comment]
Thanks so much for your help @hajimehoshi . As usual I'm amazed at your responsiveness on issues.

**Comment by hajimehoshi** [comment]
As the usage of basePixels is slightly different now, I cannot re-revert the change. I'll consider another solution to solve this. I'm sorry for the regression.

> Given that this works well on other platforms, it feels like android just doesn't handle this well. Can we make android support this pattern, or should I be doing my code differently?

Android is so special that it has to handle context losts. Then, it have to pay costs to keep pixel data. This is not your responsibility. Ebitengine should care this.

> The fix that was reverted -- did it cause a problem, or can we possibly restore it in 2.5? It seems to fix most of my problem.

The reverting was the cause, but the same solution cannot be apllied any longer. I have to invent another solution.

**Comment by hajimehoshi** [comment]
@corfe83 Please take a look at 7f3cff5e7c26ce74a80846a095f56f610cd8cb07

**Comment by corfe83** [comment]
Thanks for the response. The code change looks good to me. However, I tested and it seems very slow on my android phone, I'm only getting 1.5 FPS (consistent), and I notice GC is running 10-20 times/second, it's almost as if the allocations are worse now. I'm not sure why yet.

**Comment by corfe83** [comment]
To be clear, I was having good performance on ff926b2e9f3e , and some change between that and this has made framerate worse on android and GC much more frequent

**Comment by hajimehoshi** [comment]
I see. Would it be possible to know which commit is the cause by git-bisect?

**Comment by corfe83** [comment]
I'll try to find out.

**Comment by corfe83** [comment]
git bisect says this commit causes a performance problem on android: aca42e40467e66851dbce6e7770c7427cb029235

**Comment by corfe83** [comment]
Also, the profiling tools don't show an exact codepath, but profiler says it's clear there's some kind of problem in readPixelsFromGPU (this is the profiling data for memory allocs)

![image](https://user-images.githubusercontent.com/61067263/223647897-171cd3af-47e7-4b1a-99d1-42d68231193d.png)

**Comment by hajimehoshi** [comment]
I've removed most of extra allocations at readPixelsFromGPU.

@corfe83 Could you try 22d74e78b2b03c4b0d8b59c271c74c6a3fac4725? Thanks,

**Comment by corfe83** [comment]
Thanks @hajimehoshi . These changes are a definite improvement, I'm no longer seeing GC run multiple times per second. However, I'm still running at only 11 FPS, so I think there is still some waste in the latest 2.5. I've recorded some profiling data with the latest changes running on android, can you take a look? Unfortunately the profiling doesn't show code line numbers for ebiten, only function names, but maybe it's enough.

Zip of all profiling data:
[prof.zip](https://github.com/hajimehoshi/ebiten/files/10923421/prof.zip)

I notice that about 36%+28.5%+5% = **almost 70%** of all allocs are coming from ebiten engine, which feels like too much still:
![image](https://user-images.githubusercontent.com/61067263/223790699-d07b5c67-cac9-4f1a-b984-f84afb652a6e.png)
![image](https://user-images.githubusercontent.com/61067263/223790713-27bacfed-c12d-4b13-beca-cfadec509c15.png)
![image](https://user-images.githubusercontent.com/61067263/223790727-de3eeef0-1d1e-4bde-87d6-4de8834c19f1.png)

**Comment by hajimehoshi** [comment]
Sure, I'll take a look soon. I think I know what's going on, I am not 100% sure though...

By the way, do you read pixels from textures by `(*ebiten.Image).At` every frame?

**Comment by hajimehoshi** [comment]
Also, does this change significant compared to old version like `v2.5.0-alpha.12`?

**Comment by corfe83** [comment]
No, I only call (*ebiten.Image).At() at startup for some preprocessing. Once my game has fully initialized (within the first frame), it never calls (*ebiten.Image).At() again. I'll try comparing performance to v2.5.0-alpha.12 .

**Comment by corfe83** [comment]
v2.5.0-alpha.12 is much better! It gets me around 35-55 frames per second on android.

I still wish I could see 60 frames per second consistently, however.

**Comment by hajimehoshi** [comment]
Thanks!

I think the memory allocations at the restorable package is inevitable, as these are necessary information to restore from context lost.

@corfe83 Can I test your repository by the way?

**Comment by hajimehoshi** [comment]
Note to my self: my rough guess is that the number of draw calls changed after https://github.com/hajimehoshi/ebiten/issues/2581 is fixed.

**Comment by corfe83** [comment]
Yes, I'll give you access. One moment.

**Comment by corfe83** [comment]
I've sent you invites. Note there are two repositories, "goutil" is various utilities and ui things, and "idlearmada" is the actual game.

"bandroid" is the script I use to build the game for android from a linux environment.

**Comment by hajimehoshi** [comment]
Thank you for the invitations!

As it is too later in my location, I'll take a look tomorrow.

**Comment by corfe83** [comment]
I understand it's late, no rush.

When you are available, a couple of tips:

In the game UI, if you go to "settings" tab in the game and click "Toggle Dev Stats", you'll see a readout of FPS, GC frequency, etc. And if you want to enable profiling, Just add this line to the first line of Init() in core\main.go:

`engine.ProfilingEnabled = true`

Let the game run on android for some time with profiling enabled, and after ~3-5 minutes you should see files written here, which you can access by telling your android phone to support file transfer over USB:

`Android\data\com.musicalbox.idle.armada\files\perf`

**Comment by hajimehoshi** [comment]
@corfe83 Please try 367606f2748ae383ec87e6fcc9c7bb3636228517, thanks

**Comment by corfe83** [comment]
Thanks @hajimehoshi . I see you made a few more related fixes after 367606f , so I synced to 46173ff. I think the changes are good and the UI is responsive enough (I think I could ship with this), but I do wish the FPS were almost always 60 FPS. It seems to be 30-40 FPS on my Pixel 6 Pro.

Some tabs do reach 60 FPS, but the battle tab and the building tab are both relatively slow. I appreciate your work so far -- I wonder if there is any further improvement that can be done? I might have time to dig in and do some more profiling later.
